File: iphone/roadmap_camera.m
/* rodamap_camera.m - iphone camera implementation.
 *
 * LICENSE:
 *
 *   Copyright 2009 Avi R.
 *   Copyright 2009, Waze Ltd
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * SYNOPSYS:
 *
 *   See roadmap_camera.h
 */


#include "roadmap_main.h"
#include "roadmap_messagebox.h"
#include "roadmap_iphonemain.h"
#include "roadmap_iphonecamera.h"

void roadmap_camera_take_alert_picture(  CGSize image_size, CameraImageQuality image_quality,
								 const char* image_folder, const char* image_file )
{
   if (![RoadMapCameraView isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera])
	{
      roadmap_messagebox("Oops", "Camera is not available");
		return;
	}
   
	RoadMapCameraView *imagePicker = [[RoadMapCameraView alloc] init];
	
	[imagePicker setSourceType:UIImagePickerControllerSourceTypeCamera];
	[imagePicker takePictureWithSize: image_size
                         andQuality: image_quality
                          andFolder: image_folder
                        andFileName: image_file];
	
	[imagePicker release];
}

@implementation RoadMapCameraView

- (void) takePictureWithSize: (CGSize) image_size
					   andQuality: (CameraImageQuality) image_quality
                   andFolder: (const char*) image_folder
					  andFileName: (const char*) image_file
{
	gImageSize = image_size;
	gImageQuality = image_quality;
	gImageFolder = image_folder;
	gImageFileName = image_file;
	
	[self setDelegate:self];
	
	roadmap_main_present_modal(self);
}


//UIImagePickerController delegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingImage:(UIImage *)image editingInfo:(NSDictionary *)editingInfo
{
   CGSize newSize;
   
	//Scale image
   if (image.size.width < image.size.height) {
      newSize = gImageSize;
   } else {
      newSize.width = gImageSize.height;
      newSize.height = gImageSize.width;
   }

	
	UIGraphicsBeginImageContext( newSize );
	[image drawInRect:CGRectMake(0,0,newSize.width,newSize.height)];
	UIImage* newImage = UIGraphicsGetImageFromCurrentImageContext();
	UIGraphicsEndImageContext();
		
	//Get JPEG
	CGFloat quality;
	switch (gImageQuality) {
		case quality_low:
			quality = 0.1f;
			break;
		case quality_medium:
			quality = 0.5f;
			break;
		case quality_high:
			quality = 1.0f;
			break;
		default:
			quality = 0.5f;
			break;
	}
	
	NSData *jpgData = UIImageJPEGRepresentation(newImage, quality);
	
	//Save image
	NSString *fileName = [NSString stringWithFormat:@"%s%s", gImageFolder, gImageFileName]; //Assuming file name is with extension
	[jpgData writeToFile:fileName atomically:NO];
	
	roadmap_main_dismiss_modal();
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
	roadmap_main_dismiss_modal();
}

- (void)dealloc
{
	
	[super dealloc];
}

@end

File: android/roadmap_camera.c
/* roadmap_camera.c - Android camera API implementation
 *
 * LICENSE:
 *   Copyright 2009, Waze Ltd
 *   Alex Agranovich
 *   Ehud Shabtai
 *
 *   This file is part of RoadMap.
 *
 *   RoadMap is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License V2 as published by
 *   the Free Software Foundation.
 *
 *   RoadMap is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with RoadMap; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * SYNOPSYS:
 *
 *   See roadmap_camera.h
 */

#include <stdlib.h>
#include <string.h>
#include "roadmap_camera.h"
#include "roadmap_path.h"
#include "roadmap_camera_image.h"
#include "JNI/FreeMapJNI.h"
#include "roadmap_path.h"

// Temporary file for the image capture
static const char* gCaptureFileName = "capture_temp.jpg";

/*
 * JNI Callback context
 */
typedef struct
{
   CameraImageCaptureCallback capture_callback;
   CameraImageCaptureContext* context;
} AndrCameraCbContext;

/*
 * Only one camera capture can be done at a time. So can be assume singelton
 */
static AndrCameraCbContext sgCameraContext;

/***********************************************************
 *  Name        : roadmap_camera_take_picture
 *  Purpose     : Shows the camera preview, passes target image attributes
 *                   and defines the target capture file location
 *
 */
BOOL roadmap_camera_take_picture( CameraImageFile* image_file, CameraImageBuf* image_thumbnail )
{
    int image_quality_percent;  // Image quality in percents
    int retVal, retValThumb = -1;
    int byte_per_pix = roadmap_camera_image_bytes_pp( image_thumbnail->pixfmt );
    // Transform to percent low - 34, medium - 67, high - 100
    image_quality_percent = 34 + 33 * (int) image_file->quality;

    // Set the target path
    CAMERA_IMG_FILE_SET_PATH( *image_file, roadmap_path_user(), gCaptureFileName );

    // Call the JNI for the file
    retVal = FreeMapNativeManager_TakePicture( image_file->width, image_file->height, image_quality_percent,
            image_file->folder, image_file->file );

    // Call the JNI for the thumbnail if requested and the picture was taken successfully
    if ( ( retVal == 0 ) && image_thumbnail )
    {
        // Assume 4 byte per pixel
        // Only int array can be accepted for the Android
        // Malloc SHOULD preserve 4-byte alignment so further casting is possible !!!!!!!
        image_thumbnail->buf = malloc( image_thumbnail->width * image_thumbnail->height * byte_per_pix );
        retValThumb = FreeMapNativeManager_GetThumbnail( image_thumbnail->width, image_thumbnail->height,
                                                                    byte_per_pix, (int*) image_thumbnail->buf );
        if ( retValThumb != 0 )
        {
            roadmap_log( ROADMAP_WARNING, "Thumbnail request to Android is failed" );
        }
    }
    return ( retVal == 0 );
}



/***********************************************************
 *  Name        : roadmap_camera_take_picture_async
 *  Purpose     : Asynchronous function. Main thread still running.
 *                Shows the camera preview, passes target image attributes
 *                   and defines the target capture file location
 *
 */
BOOL roadmap_camera_take_picture_async( CameraImageCaptureCallback callback, CameraImageCaptureContext* context )
{
    int image_quality_percent;  // Image quality in percents
    int retVal, retValThumb = -1;
    CameraImageBuf* image_thumbnail = &context->image_thumbnail;
    CameraImageFile* image_file = &context->image_file;
    // Transform to percent low - 34, medium - 67, high - 100
    image_quality_percent = 34 + 33 * (int) image_file->quality;

    // Set the target path
    CAMERA_IMG_FILE_SET_PATH( *image_file, roadmap_path_user(), gCaptureFileName );

    // Call the JNI for the file
    retVal = FreeMapNativeManager_TakePictureAsync( image_file->width, image_file->height, image_quality_percent,
            image_file->folder, image_file->file );

    sgCameraContext.context = context;
    sgCameraContext.capture_callback = callback;

    return TRUE;
}


void roadmap_camera_take_picture_async_cb( int res )
{
   int retValThumb = -1;

   // Call the JNI for the thumbnail if requested and the picture was taken successfully
   if ( res )
   {
      CameraImageCaptureContext* context = sgCameraContext.context;
      CameraImageBuf* image_thumbnail = &context->image_thumbnail;
      int byte_per_pix = roadmap_camera_image_bytes_pp( image_thumbnail->pixfmt );
     // Assume 4 byte per pixel
     // Only int array can be accepted for the Android
     // Malloc SHOULD preserve 4-byte alignment so further casting is possible !!!!!!!
     image_thumbnail->buf = malloc( image_thumbnail->width * image_thumbnail->height * byte_per_pix );
     retValThumb = FreeMapNativeManager_GetThumbnail( image_thumbnail->width, image_thumbnail->height,
                                                                 byte_per_pix, (int*) image_thumbnail->buf );
     if ( retValThumb != 0 )
     {
         roadmap_log( ROADMAP_WARNING, "Thumbnail request to Android is failed" );
     }
   }
   sgCameraContext.capture_callback( sgCameraContext.context, res );
}

